### PR TITLE
SAM GMAC ethernet driver fixes

### LINKED ISSUE #11255: Pinging sam_e70 leads to unresponsive ethernet device at some point
On target a net ping shows it can send/recv packets.

From an external host: ping -i 0.00001 <ip of the sam_e70 board> will work for some time until it cannot reach the target. On target side, a net ping does not work anymore.

Seems like either the ethernet driver or the ethernet device on the sam_e70 gets unresponsive.

### PR DESCRIPTION
**Pull request by aurel32**
This set of 3 patches aims to fix issue #11255. It has been tested with ICMPv4 and ICMPv6 ping with a 0.00001 interval during 2 hours, both as standalone and with PR#11888 + PR#12555. For me it fixes the reported issue, although I am almost sure there are more bugs lying down in the driver, especially as these patches only touch the TX path.

Fixes: #11255

### CODE REVIEW

**Review by jukkar — COMMENTED** [review]
Thanks @aurel32, the commit messages in this PR are excellent :+1: 
I will try this with gptp which had some weird issues with sam-e70 earlier. I will post results after some testing.

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
Done some testing with gptp, sam-e70 works now as well as frdm-k64f

**Comment by jukkar** [comment]
Now I started to see this in sam-e70
```
[00:00:04.546,287] <wrn> net_gptp: Not AS capable: 4294967295 ns > 100000 ns
```
This is most probably not related to this PR but anyway this is something I do not see in frdm-k64f. This needs more investigation in gptp code.

File: drivers/ethernet/eth_sam_gmac_priv.h [hunk 1]
```diff
@@ -31,7 +31,8 @@
 /** RX descriptors count for main queue */
 #define MAIN_QUEUE_RX_DESC_COUNT CONFIG_ETH_SAM_GMAC_BUF_RX_COUNT
 /** TX descriptors count for main queue */
-#define MAIN_QUEUE_TX_DESC_COUNT (CONFIG_NET_BUF_TX_COUNT + 1)
+#define MAIN_QUEUE_TX_DESC_COUNT (max(CONFIG_NET_BUF_RX_COUNT, \
```
Comment (pfalcon) [inline drivers/ethernet/eth_sam_gmac_priv.h]: Please add comment why this formula.

**Review by pfalcon — CHANGES_REQUESTED** [review]

File: drivers/ethernet/eth_sam_gmac_priv.h (hunk 1 above)
Comment (pfalcon) [reply drivers/ethernet/eth_sam_gmac_priv.h]: While doing that, also please update the title from "drivers: eth: gmac: correctly compute the number of descriptors" to "drivers: eth: gmac: *conservatively* compute the number of descriptors". Per, https://github.com/zephyrproject-rtos/zephyr/pull/12591#issuecomment-456035299 , this is unlikely "correct".

Comment (pfalcon) [reply drivers/ethernet/eth_sam_gmac_priv.h]: Also:

> This wrong assume that 

Should be "wrongly assume".

**Comment by pfalcon** [comment]
> The ICMPv6 implementation already copies the RX buffer into the TX buffers.

Well, good to know (I didn't).

> Should I try to do the same for ICMPv4 instead of this patch?

"Instead", unlikely, at least I wouldn't burden anyone with such a request. But if you could pursue that after this patch, I would both appreciate and +1 (well, +10) it. I generally mention such issues to "raise awareness" and build a consensus on how to resolve them.

**Comment by aurel32** [comment]
> > Should I try to do the same for ICMPv4 instead of this patch?
> 
> "Instead", unlikely, at least I wouldn't burden anyone with such a request. But if you could pursue that after this patch, I would both appreciate and +1 (well, +10) it. I generally mention such issues to "raise awareness" and build a consensus on how to resolve them.

Ok, I'll keep that on my TODO list, as it would also allow to have a simple TX timeout mechanism in the GMAC driver.

**Comment by aurel32** [comment]
@pfalcon I have just pushed a new version with the fixes you suggested.

**Review by pfalcon — APPROVED** [review]
Thanks for updates!

**Review by mnkp — COMMENTED** [review]
@aurel32 I didn't have time to review this PR in a timely manner. Just want to say thanks for working on and fixing the issues!

**Pull request merged**